### PR TITLE
[SPARK-25815][k8s] Support kerberos in client mode, keytab-based token renewal.

### PR DESCRIPTION
**Pull request by vanzin**
This change hooks up the k8s backed to the updated HadoopDelegationTokenManager,
so that delegation tokens are also available in client mode, and keytab-based token
renewal is enabled.

The change re-works the k8s feature steps related to kerberos so
that the driver does all the credential management and provides all
the needed information to executors - so nothing needs to be added
to executor pods. This also makes cluster mode behave a lot more
similarly to client mode, since no driver-related config steps are run
in the latter case.

The main two things that don't need to happen in executors anymore are:

- adding the Hadoop config to the executor pods: this is not needed
  since the Spark driver will serialize the Hadoop config and send
  it to executors when running tasks.

- mounting the kerberos config file in the executor pods: this is
  not needed once you remove the above. The Hadoop conf sent by
  the driver with the tasks is already resolved (i.e. has all the
  kerberos names properly defined), so executors do not need access
  to the kerberos realm information anymore.

The change also avoids creating delegation tokens unnecessarily.
This means that they'll only be created if a secret with tokens
was not provided, and if a keytab is not provided. In either of
those cases, the driver code will handle delegation tokens: in
cluster mode by creating a secret and stashing them, in client
mode by using existing mechanisms to send DTs to executors.

One last feature: the change also allows defining a keytab with
a "local:" URI. This is supported in client mode (although that's
the same as not saying "local:"), and in k8s cluster mode. This
allows the keytab to be mounted onto the image from a pre-existing
secret, for example.

Finally, the new code always sets SPARK_USER in the driver and
executor pods. This is in line with how other resource managers
behave: the submitting user reflects which user will access
Hadoop services in the app. (With kerberos, that's overridden
by the logged in user.) That user is unrelated to the OS user
the app is running as inside the containers.

Tested:
- client and cluster mode with kinit
- cluster mode with keytab
- cluster mode with local: keytab
- YARN cluster with keytab (to make sure it isn't broken)

### CODE REVIEW

**Comment by mccheah** [comment]
@ifilonenko @skonto

**Comment by ifilonenko** [comment]
Just a note, this PR has some overlaps with https://github.com/apache/spark/pull/22915. But just after a brief skim, I am a bit hesitant on storing keytabs in secrets, but it might have to be done for client mode support.

**Comment by ifilonenko** [comment]
> But whether to store it in secrets is not a question. You either store it in a secret or you don't support the keytab/principal feature in Spark at all, and we can delete a bunch of code here.

But can we, for a specific mode, client or cluster, tune whether the keytab is put into a secret? Right now, if a keytab is specified, regardless, a secret is created. I do wish to support keytab/principal so I think that storage in secrets should be fine.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala
Comment (rvesse) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: Typo - `Wit` -> `With`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala
Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala]: I'm not sure I follow your train of thought here so I'll comment on what I understand.

First, the code that creates the secret is in `KerberosConfDriverFeatureStep`. As far as I know, that class is not used in client mode. In client mode the keytab stays in the client machine, with the driver, and the driver just sends DTs to executors. So the whole discussion about secrets is irrelevant in that case.

In cluster mode, you need the driver to have access to the keytab for this feature to work. There are a few ways to achieve that:

- the current YARN mode, which is the keytab lives in the submission host, and is distributed with the application. In k8s this would amount to what I have here: the submission code creates a secret for the driver pod and stashes the keytab in it.

- add the ability to store the keytab in an external place (like HDFS or an HTTP server). That has drawbacks (e.g. people probably wouldn't like that, and there's a chicken & egg problem in HDFS, so you'd still need a kerberos TGT to bootstrap things).

- add a k8s-specific feature of mounting a pre-defined secret in the driver pod. I believe this is what you're suggesting?

I think supporting the first is easy as this change shows, and keeps feature parity with what's already supported in YARN. Unless there's a glaring issue with using secrets that I'm not aware of, I don't see a reason for not doing it.

The third option (pre-defined secret) could also be added. My hope is that you could do it with pre-existing configs (`spark.kubernetes.driver.secrets.` & company), but I don't know how you'd set the `spark.kerberos.keytab` and `spark.kerberos.principal` configs just in the driver - and not in the submission client. So it seems we'd need at least a little bit of code here to support that scenario.

Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala]: Ah, ok I get it now. I can do that. I'll try to include support for (3) but it depends on how much I have to touch other parts of the code. Hopefully not much.

Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala]: Yeah, I can always throw up a follow-up for that. No worries

**Comment by vanzin** [comment]
To address the comments I had to change a whole bunch of things; mainly because I noticed that the client and cluster paths were really different, when they should for the most part be similar.  Check the updated PR description for details.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkPod.scala
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkPod.scala]: I added this because I started to get tired of code like the following:

```
val someIntermediateName = someOption.map { blah =>
   // create the updated pod
}.getOrElse(previousPodName)

// lather, rinse, repeat
```

To me that's hard to follow and brittle, and this pattern makes things clearer IMO.

Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkPod.scala]: I would think that this change is out of the scope of this PR, but I do love the use of a PartialFunction here. Thanks for this!

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DelegationTokenFeatureStep.scala
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DelegationTokenFeatureStep.scala]: During the recent discussion in the sig meeting I was thinking about this guy. I think it may not be necessary to mount this secret in executors, and let the driver propagate the tokens to executors through its normal means.

e.g. later if/when adding code to monitor delegation tokens for updates, that would mean only the driver has to do it.

I'll git this a try; if it works, it might be feasible to re-merge this code with the kerberos step.

Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DelegationTokenFeatureStep.scala]: This is what I meant above when I said that the `HadoopKerberosLogin` logic could be deleted. The assumption here is that the secret should not be created as the keytab will use the HadoopDelegationTokenManager logic. The only secret that should be _created_ would be the keytab. However, I personally thought that we should point to a secretName that is either the delegationToken or the keytab. Hence why I suggested that the secretName and secretItemKey remain.

Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DelegationTokenFeatureStep.scala]: I think you're missing one use case there. There are 3 use cases:

- keytab: keytab is provided to driver, driver handles kerberos login, creates tokens, distributes tokens to executors
- pre-defined token secret: secret is mounted in the driver, env variable is set, driver loads them and distributes to executors
- default kerberos case: submission client generates delegation tokens, creates a secret for them, then this behaves like the bullet above

The third use case is actually the most common. In your reply above you're only covering the other two. My code covers all three.

I'm just saying that this code actually doesn't need to do anything on the executor side, because the driver takes care of everything when the credentials are provided.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala]: I see that you noted that this is always done across resource managers. What is the reason for that, just wondering? as I introduced it exclusively in the HadoopSteps

Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala]: If you don't do this, whatever is the OS user in the container will become the identity used to talk to Hadoop services (when kerberos is not on).

In YARN, for example, that would be the "yarn" user.

In k8s, with the current image, that would be "root".

You probably don't want that by default. We're talking about non-secured Hadoop here, so users can easily override this stuff, but by default let's at least try to identify the user correctly.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: nit: Why isn't this logic done in a `KubernetesDelegationTokenManager` or some Util class? It seems we are over-running this step with a lot of functionality? no? You are also just mocking via
`when(step).createDelegationToken()`, I personally don't agree with the coverage brought up by this method of unit-testing but this is an extension of the discussion on a separate PR. i.e. I would mock the tokenManager and test to ensure that the `obtainDelegationTokens` take in a particular `creds` as an argument (also, so that one could test the if statement below).`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: nit: is there a particular advantage gained by this style of Seq() building? I prefer the .toSeq approach as I personally don't like `if {...} else { Nil }`. Once again, why was the `Util` class removed? Isn't this a lot of logic to throw into this one step?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: nit: `otherStep` isn't very descriptive to the contents of the steps. I prefer `hadoopSteps` if anything

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala
Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: > Why isn't this logic done in a KubernetesDelegationTokenManager or some Util class?

Because that's not needed. What exactly would doing that achieve?

> I would mock the tokenManager and test to ensure that the obtainDelegationTokens take in a particular creds

That's not what this code is testing. This code is testing that when delegation tokens are created, they are stashed in a secret. For that I'm mocking the creation of delegation tokens, because without any services to talk to, delegation tokens will not be created.

(BTW, that's what mocks should be used for.)

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala]: Should we check the contents of the token in the unit test?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala
Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: re: the style, same reasoning as the `transform` method I added. It avoids having to create a bunch of local variables and then concatenate them at the end.

re: the Util class, it existed because the methods were being called from multiple feature steps. Now there's only one step that does that stuff, so the util class became unnecessary.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala
Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: All this code will go away when I get to SPARK-25877, so I'm not super concerned with the names here.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala
Comment (vanzin) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala]: Seems redundant, but why not.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: > the style, same reasoning as the transform method I added. It avoids having to create a bunch of local variables and then concatenate them at the end.

The `transform` I like because it gets rid of: 
```
  val someIntermediateName = someOption.map { blah =>
     // create the updated pod
  }.getOrElse(previousPodName)
```
But that is not what is happening here exactly. Personally, I thought that the previous approach (with local variables + concatonation) was quite clear and didn't really require a change. But its a matter of opinion.

> Now there's only one step that does that stuff, so the util class became unnecessary.

Okay I guess from a re-usability perspective that is true, but imo seems like a lot of logic going into this step. but once again, just a matter of opinion

Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: > seems like a lot of logic going into this step

The logic itself isn't really different from before. The only difference is that the "what is done" for each combination of configs now also lives here.

It could be broken down into separate steps, but then you'd have 3 related steps that need to agree on how to treat the configuration of the driver; e.g. if the step that adds they keytab runs, then the step that creates delegation tokens shouldn't run. And that makes the code harder to follow, because you have to look at multiple places to understand how this works.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala
Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: > I'm mocking the creation of delegation tokens

BTW there are other ways to achieve that, e.g. with `UserGroupInformation.createTestUser()` and `doAs`, but I think the current test is easier to understand.

Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: > Because that's not needed. What exactly would doing that achieve?

I mean that is the function of a DelegationTokenManager, no? To handle nuances related to DT management, both for creation and renewal. _shrug_

> This code is testing that when delegation tokens are created, they are stashed in a secret

Well there is logic being introduced here beyond just `tokenManager.obtainDelegationToken()`, right? And if the purpose of the tests is to take the results from delegation token creation and put into a secret, then it would make sense to mock a `KubernetesDelegtionTokenManager` and just call `kdtm.obtainTokens()`, where this logic would be housed. Else, if we were to keep this logic within this step, i'd prefer that all parts of the logic are mocked. 

Once again, a matter of opinion, that differs between unit-testing strategy. I'll loop in @mccheah

> I think the current test is easier to understand.

I agree, but is it comprehensive is what I am talking about here?

Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: > that is the function of a DelegationTokenManager. To handle nuances related to DT management

Yes. And the test here is not interested in whether DT creation works or not. It's interested in that if they are created, then they are stashed in a secret.

> then it would make sense to mock a KubernetesDelegtionTokenManager and just call kdtm.obtainTokens()

Would it make you happier if instead of having instead of the current method, have this:

```
def createDelegationTokens(creds: Credentials)
```

The only extra thing you'd really be testing is the call to `SparkHadoopUtil.get.serialize`. I don't really see a lot of gains there.

Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala]: > Hmm, I somewhat disagree, but maybe I am just overly opinionated.

The code that creates delegation tokens is not in this class. If delegation token creation is to be tested, it needs to be tested in the test code for `HadoopDelegationTokenManager`.

All this class does is take the tokens that other class creates and stash them in a secret.

I'll change this to not need a mock at all, mostly to make a point.

**Comment by suxingfate** [comment]
--keytab is a simpler way IMO to enable long running spark on k8s than DT renew approach

**Comment by vanzin** [comment]
retest this please

**Comment by shaneknapp** [comment]
test this please

**Comment by vanzin** [comment]
there was a seemingly corrupt xml file in the jenkins worker, I removed it and will retest.

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
on a non-testing not, any further feedback here?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala]: The reason I am particular about using mocks is in the case that the internal logic changes (i.e. we remove using `SparkHadoopUtil` in future iterations, or whatnot.) But tbh, I am not hard-set on this as I see your point, and such changes aren't probable surrounding this code. So I am guess I am fine with it.

**Review by ifilonenko — APPROVED** [review]
This LGTM pending tests passing and a +1 from @mccheah.
I would also love to see this in as soon as possible as the features contained here are extremely desirable.

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
No more comments here and I plan to push this tomorrow based on the +1s I have.

**Comment by erictcgs** [comment]
Hi @vanzin , great work!  I'd just been trying to get this working and was happy to find your PR.  I'm having some trouble getting it working, though - not sure whether an issue in my env or if there's a condition that's not getting caught, hoping you or others could shed some light:

Summary:
- Executors don't seem to be logging in under any scenario using the delegation token
- In client mode, driver connects to hdfs and gets delegation token but the executors don't seem to be using it
- In cluster mode, seems driver pod doesn't get hadoop configs mapped in (this had been working in previous version) so doesn't try using kerberos keytab.  Once I manually add the hadoop configs the driver is able to connect to hdfs, but the executors fail

Environment overview:
- Kerberized HDFS in traditional hadoop cluster
- Launching spark into separate kubernetes cluster, trying to access a parquet dataset

Steps:
1. Cloned your branch, commit cce3f1dd32c204f5c57b3db3a79d49895512f0a5 from today
1. Compiled, created docker images
1. kinit-ed
1. Tried running a small python program to access hdfs, launching in local, cluster, and client modes

**Results - Local mode**

```
./bin/spark-submit \
--master local[4] \
--name spark-hdfs \
--principal user@XXX.COM \
--keytab keytab.kt \
https://xxx.com/spark/hdfs_test.py
```

works, program runs and is able to load parquet and print the first row

**Results - Client mode**

```
./bin/spark-submit \
--master k8s://https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT \
--deploy-mode client \
--conf spark.driver.host=$(getent hosts $HOSTNAME | cut -f1 -d' ') \
<pod-spec-details>
--conf spark.kubernetes.kerberos.krb5.path=/path/krb5.conf \
--conf spark.kerberos.keytab=keytab.kt \
--conf spark.kerberos.principal=user@XXX.COM \
https://xxx.com/spark/hdfs_test.py
```

Driver is able to login to the hdfs server, acquires a delegation token, starts executors

Executors go to connect to HDFS, error out with the following:
```
2018-12-14 00:13:41 WARN  Client:683 - Exception encountered while connecting to the server : org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
2018-12-14 00:13:41 ERROR Executor:95 - Exception in task 0.3 in stage 0.0 (TID 3)
org.apache.spark.SparkException: Exception thrown in awaitResult:
        at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:226)
        at org.apache.spark.util.ThreadUtils$.parmap(ThreadUtils.scala:290)
...
Caused by: java.io.IOException: Failed on local exception: java.io.IOException: org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]; Host Details : local host is: "hdfs-test-1544746404469-exec-1/10.244.42.225"; destination host is: "xxx.com":5555;
        at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:776)
        at org.apache.hadoop.ipc.Client.call(Client.java:1480)
        at org.apache.hadoop.ipc.Client.call(Client.java:1413)
```

Seems like they're not trying to connect with the delegation token?

**Results - Cluster mode**

```
./bin/spark-submit \
--master k8s://https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT \
--deploy-mode cluster \
<pod-spec-details>
--conf spark.kubernetes.kerberos.krb5.path=/path/krb5.conf \
--conf spark.kerberos.keytab=keytab.kt \
--conf spark.kerberos.principal=user@XXX.COM \
https://xxx.com/spark/hdfs_test.py
```

Driver and executor pods are launched

Driver tries to connect to hdfs with SIMPLE authentication

Found that hadoop configs weren't being propagated to driver pod, even when adding the `--conf spark.kubernetes.hadoop.configMapName` option

When I manually added the configs and set the driver environment variable HADOOP_CONF_DIR, driver now is able to connect to hdfs and get delegation token, but executors have the same problem observed in client mode

**Comment by vanzin** [comment]
I've been running tests with the Spark `HdfsTest` example; maybe pyspark does something different. I'll try that tomorrow.

I haven't really tried the config map route, only `HADOOP_CONF_DIR`.

**Comment by rvesse** [comment]
This might be a bug, I've complained in the past that HDFS stuff doesn't seem to work unless you set `HADOOP_CONF_DIR` in the submission clients environments

**Comment by vanzin** [comment]
This one might have been just a bad merge on my part. Without integration tests this is hard to catch (and I was having trouble with minikube yesterday).

I'll try out my tests on minikube tomorrow.

**Comment by erictcgs** [comment]
Had a chance to try it - the driver pod is now getting the keytab mounted and is able to generate delegation token, so that's looking good.  I'm still having an issue where the executors seem to not be using the dt and so are getting denied access - are they also supposed to have hadoop_conf_dir set?

**Comment by erictcgs** [comment]
@vanzin - how are you compiling spark?  Specifically, what version of hadoop are you testing with?  I'm working with v2.7.7, wondering if there are changes in later versions to the ipc.client and usergroupinformation classes that may handle delegation tokens differently

**Comment by vanzin** [comment]
I've tried this again in all configurations I know (and also pyspark) and it's working for me:

- cluster mode: keytab, local: keytab, no keytab
- client mode: keytab / no keytab, client in my laptop, client in k8s pod
- spark-shell and pyspark shell with tgt

I'm building Spark normally. Make sure your images are updated in docker (remember that the pyspark image is *not* build by default anymore).

You should see a log message like this in the executors:

```
2018-12-14 18:52:00 INFO  SparkHadoopUtil:58 - Updating delegation tokens for current user.
```

**Comment by erictcgs** [comment]
Update - my setAuthenticationMethod change didn't work - after calling that I see that the executor reports it's trying to use TOKEN method, but it's still getting rejected with an error that it can't login via TOKEN, KERBEROS:
```
18/12/14 23:26:01 DEBUG UserGroupInformation: PrivilegedActionException as:user (auth:TOKEN) cause:org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
18/12/14 23:26:01 DEBUG UserGroupInformation: PrivilegedAction as:user (auth:TOKEN) from:org.apache.hadoop.ipc.Client$Connection.handleSaslConnectionFailure(Client.java:651)
18/12/14 23:26:01 WARN Client: Exception encountered while connecting to the server : org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
```

As a test I tried mounting in my keytab and hadoop_conf_dir into the executors and setting KRB5CCNAME env var (via pod template) - this results in the main user being kerberized, then I commented out the "runassparkuser" section of the coarsescheduler, and now the executors are able to successfully login to HDFS and grab the parquet data.  

It looks like there's some incompatibility in my environment with how the delegation tokens are being used - I think this shouldn't block this merge if it's working for others - is there any other venue people would suggest I continue trying to debug?

**Comment by vanzin** [comment]
I suggest filing a bug or starting a mailing list thread. But really I doubt anyone will be able to help if you only provide the errors, without explaining your environment and, especially, the code you're running.

**Comment by vanzin** [comment]
retest this please

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStep.scala
Comment (turboFei) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStep.scala:40]: gentle ping @vanzin 

After this pr, seems the HADOOP_CONFIG_MAP is not mounted on executor side.

how does the executor get hadoop conf?

**Comment by turboFei** [comment]
> The main two things that don't need to happen in executors anymore are:
> adding the Hadoop config to the executor pods: this is not needed
> since the Spark driver will serialize the Hadoop config and send
> it to executors when running tasks.

gentle ping @vanzin 
seems the executor still need hadoop config.
image

**Pull request closed without merging**
